### PR TITLE
dialect: (riscv) Custom print attributes

### PR DESCRIPTION
**Pull request by kingiler**

### CODE REVIEW

File: xdsl/dialects/riscv.py [hunk 1]
```diff
@@ -351,15 +350,45 @@ def parse(cls, parser: Parser) -> Self:
...
+                operand := parser.parse_optional_unresolved_operand()
+            ):
+                operands.append(operand)
+            return operands
+        return []
+
+    @classmethod
+    def parse_attributes(cls, parser: Parser) -> Mapping[str, Attribute]:
+        """
+        Parse custom printed attributes. Subclass may overwrite this method.
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:372]:
```suggestion
        Parse attributes with custom syntax. Subclasses may override this method.
```

File: xdsl/dialects/riscv.py [hunk 2]
```diff
@@ -351,15 +350,45 @@ def parse(cls, parser: Parser) -> Self:
...
         if self.operands:
             printer.print(" ")
             printer.print_list(self.operands, printer.print_operand)
-        printer.print_op_attributes(self.attributes)
+        self.print_attributes(printer)
         printer.print_regions(self.regions)
         printer.print(" : ")
         printer.print_operation_type(self)
 
+    def print_attributes(self, printer: Printer) -> Set[str]:
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:385]:
```suggestion
    def custom_print_attributes(self, printer: Printer) -> Set[str]:
```

File: xdsl/dialects/riscv.py [hunk 3]
```diff
@@ -351,15 +350,45 @@ def parse(cls, parser: Parser) -> Self:
...
+        """
+        Custom print specific attributes and return the set of custom printed attributes. Subclass may overwrite this method.
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:387]:
```suggestion
        Print attributes with custom syntax. Return the names of the attributes printed. Subclasses may override this method.
```

**Review by superlopuh — APPROVED** [review]

File: xdsl/dialects/riscv.py [hunk 4]
```diff
@@ -333,13 +334,11 @@ def assembly_line(self) -> str | None:
...
-            parser.parse_optional_unresolved_operand,
-            parser.parse_unresolved_operand,
-        )
-        if args is None:
-            args = []
-        attributes = parser.parse_optional_attr_dict()
+        args = cls.parse_unresolved_operand(parser)
+        custom_attributes = cls.parse_attributes(parser)
+        remaining_attributes = parser.parse_optional_attr_dict()
+        # TODO ensure distinct keys for attributes
```
[the remark this answers is not in the document]
Comment (kingiler) [reply xdsl/dialects/riscv.py:340]: Filed an issue #1369

**Review by adutilleul — APPROVED** [review]

**Pull request merged**